File: WQChartsDemo/Objc/ViewController/RadarChartVC.m
// https://github.com/CoderWQYao/WQCharts-iOS
//
// RadarChartVC.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "RadarChartVC.h"

@interface RadarChartVC () 

@property (nonatomic, strong) WQRadarChartView* chartView;
@property (nonatomic, strong) NSMutableArray<WQRadarChartPolygon*>* items;
@property (nonatomic, strong) NSArray<UIColor*>* colors;

@end

@implementation RadarChartVC

- (NSArray<UIColor *> *)colors {
    if(!_colors) {
        _colors = Colors;
    }
    return _colors;
}

- (UIView *)createChartView {
    _chartView = [[WQRadarChartView alloc] init];
    _chartView.chart.dataSource = self;
    return _chartView;
}

- (void)configChartOptions {
    [super configChartOptions];
    
    __weak typeof(self) weakSelf = self;
    
    [self.optionsView addItem:ListCell.new
     .setTitle(@"IndicatorCount")
     .addItem(SliderCell.new
              .setSliderValue(0,10,5)
              .setOnValueChange(^(SliderCell* cell,float value) {
        WQRadarChartView* chartView = weakSelf.chartView;
        [chartView.chart setNeedsReloadIndicators];
        [chartView redraw];
    }))];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"IndicatorsLine")
     .setOptions(@[@"OFF",@"ON",@"Dash"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQRadarChartView* chartView = weakSelf.chartView;
        [chartView.chart setNeedsReloadIndicators];
        [chartView redraw];
    })];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"IndicatorsText")
     .setOptions(@[@"OFF",@"ON"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQRadarChartView* chartView = weakSelf.chartView;
        [chartView.chart setNeedsReloadIndicators];
        [chartView redraw];
    })];
    
    [self.optionsView addItem:ListCell.new
     .setTitle(@"SegmentCount")
     .addItem(SliderCell.new
              .setSliderValue(0,10,5)
              .setOnValueChange(^(SliderCell* cell,float value) {
        WQRadarChartView* chartView = weakSelf.chartView;
        [chartView.chart setNeedsReloadSegments];
        [chartView redraw];
    }))];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"SegmentsShape")
     .setOptions(@[@"Polygon",@"Arc"])
     .setSelection(0)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQRadarChartView* chartView = weakSelf.chartView;
        [chartView.chart setNeedsReloadSegments];
        [chartView redraw];
    })];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"SegmentsLine")
     .setOptions(@[@"OFF",@"ON",@"Dash"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQRadarChartView* chartView = weakSelf.chartView;
        [chartView.chart setNeedsReloadSegments];
        [chartView redraw];
    })];
    
}

#pragma mark - Items

- (void)configChartItemsOptions {
    [super configChartItemsOptions];
}

- (NSMutableArray<WQRadarChartPolygon*>*)items {
    if(!_items) {
        NSMutableArray<WQRadarChartPolygon*>* items = [NSMutableArray array];
        for (NSInteger i=0; i<2; i++) {
            WQRadarChartPolygon* item = [self createItemAtIndex:i];
            if (item) {
                [items addObject:item];
            }
        }
        _items = items;
    }
    return _items;
}

- (NSString*)itemsOptionTitle {
    return @"Polygons";
}

- (WQRadarChartPolygon*)createItemAtIndex:(NSInteger)index {
    if (index >= self.colors.count) {
        return nil;
    }
    UIColor* color = self.colors[index];
    WQPolygonChart* chart = [[WQPolygonChart alloc] init];
    chart.paint.fill.color = [color colorWithAlphaComponent:0.5];
    chart.paint.stroke = nil;
    return [[WQRadarChartPolygon alloc] initWithChart:chart];
}

- (SectionCell*)createItemCellWithItem:(WQRadarChartPolygon*)item atIndex:(NSInteger)index {
    __weak typeof(self) weakSelf = self;
    return SectionCell.new
    .setObject(item)
    .setTitle([NSString stringWithFormat:@"Polygon%ld",index])
    .setOnReload(^(SectionCell* cell) {
        WQRadarChartPolygon* polygon = (WQRadarChartPolygon*)cell.objcect;
        polygon.needsReloadItems = YES;
        [weakSelf.chartView redraw];
    });
}

- (void)itemsDidChange:(NSMutableArray*)items {
    [self.chartView.chart setNeedsReloadPolygons];
    [self.chartView redraw];
}

#pragma mark - RadarChartDataSource

- (NSInteger)numberOfIndicatorsInRadarChart:(WQRadarChart *)radarChart {
    return [self sliderIntegerValueForKey:@"IndicatorCount" atIndex:0];
}

- (WQRadarChartIndicator *)radarChart:(WQRadarChart *)radarChart indicatorAtIndex:(NSInteger)index {
    WQRadarChartIndicator* indicator = [[WQRadarChartIndicator alloc] init];
    [self setupStrokePaint:indicator.paint color:Color_White type:[self radioCellSelectionForKey:@"IndicatorsLine"]];
    
    if([self radioCellSelectionForKey:@"IndicatorsText"] != 0) {
        WQChartText* text = [[WQChartText alloc] init];
        text.font = [UIFont systemFontOfSize:11];
        text.color = Color_White;
        text.string = [NSString stringWithFormat:@"I-%ld",index];
        
        WQChartTextBlocks* textBlocks = WQChartTextBlocks.new;
        textBlocks.offsetByAngle = ^CGFloat(WQChartText * _Nonnull chartText, CGSize size, CGFloat angle) {
            return 15;
        };
        text.delegateUsingBlocks = textBlocks;
        
        indicator.text = text;
    }
    
    return indicator;
}

- (NSInteger)numberOfSegmentsInRadarChart:(WQRadarChart *)radarChart {
    return [self sliderIntegerValueForKey:@"SegmentCount" atIndex:0];
}

- (WQRadarChartSegment *)radarChart:(WQRadarChart *)radarChart segmentAtIndex:(NSInteger)index {
    WQRadarChartSegment* segment = [[WQRadarChartSegment alloc] init];
    [self setupStrokePaint:segment.paint color:Color_White type:[self radioCellSelectionForKey:@"SegmentsLine"]];
    segment.shape = [self radioCellSelectionForKey:@"SegmentsShape"] != 0 ? WQRadarChartSegmentShapeArc : WQRadarChartSegmentShapePolygon;
    segment.value = 1.0 - 1.0 / [self numberOfSegmentsInRadarChart:radarChart] * index;
    return segment;
}

- (NSInteger)numberOfPolygonsInRadarChart:(WQRadarChart *)radarChart {
    return self.items.count;
}

- (WQRadarChartPolygon *)radarChart:(WQRadarChart *)radarChart polygontAtIndex:(NSInteger)index {
    return self.items[index];
}

- (WQPolygonChartItem *)radarChart:(WQRadarChart *)radarChart itemOfChartForPolygonAtIndexPath:(NSIndexPath *)indexPath {
    WQPolygonChartItem* item = [[WQPolygonChartItem alloc] initWithValue:[NSNumber randomCGFloatFrom:0 to:1]];
    item.axisPaint = nil;
    return item;
}

#pragma mark - Animation

- (void)appendAnimationKeys:(NSMutableArray<NSString *> *)animationKeys {
    [super appendAnimationKeys:animationKeys];
    [animationKeys addObject:@"Values"];
}

- (void)prepareAnimationOfChartViewForKeys:(NSArray<NSString*>*)keys {
    [super prepareAnimationOfChartViewForKeys:keys];
    
    if ([keys containsObject:@"Values"]) {
        for (WQRadarChartPolygon* polygon in self.items) {
            for (WQPolygonChartItem* item in polygon.chart.items) {
                item.valueTween = [[WQChartCGFloatTween alloc] initWithFrom:item.value to:[NSNumber randomCGFloatFrom:0 to:1]	];
            }
        }
    }
    
}

@end

File: WQChartsDemo/Swift/ViewController/RadarChartVC.swift
// https://github.com/CoderWQYao/WQCharts-iOS
//
// RadarChartVC.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class RadarChartVC: RadialChartVC<RadarChartView>, ItemsOptionsDelegate, RadarChartDataSource  {
    
    lazy var colors: [UIColor] = {
        return Colors
    }()
    
    override func chartViewDidCreate(_ chartView: RadarChartView) {
        super.chartViewDidCreate(chartView)
        chartView.chart.dataSource = self
    }
    
    override func configChartOptions() {
        super.configChartOptions()
        
        optionsView.addItem(ListCell()
            .setTitle("IndicatorCount")
            .addItem(SliderCell()
                .setSliderValue(0,10,5)
                .setOnValueChange({[weak self](cell, value) in
                    guard let self = self else {
                        return
                    }
                    let chartView = self.chartView
                    chartView.chart.setNeedsReloadIndicators()
                    chartView.redraw()
                })
            )
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("IndicatorsLine")
            .setOptions(["OFF","ON","Dash"])
            .setSelection(1)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                chartView.chart.setNeedsReloadIndicators()
                chartView.redraw()
            })
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("IndicatorsText")
            .setOptions(["OFF","ON"])
            .setSelection(1)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                chartView.chart.setNeedsReloadIndicators()
                chartView.redraw()
            })
        )
        
        
        optionsView.addItem(ListCell()
            .setTitle("SegmentCount")
            .addItem(SliderCell()
                .setSliderValue(0,10,5)
                .setOnValueChange({[weak self](cell, value) in
                    guard let self = self else {
                        return
                    }
                    let chartView = self.chartView
                    chartView.chart.setNeedsReloadSegments()
                    chartView.redraw()
                })
            )
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("SegmentsShape")
            .setOptions(["Polygon","Arc"])
            .setSelection(0)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                chartView.chart.setNeedsReloadSegments()
                chartView.redraw()
            })
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("SegmentsLine")
            .setOptions(["OFF","ON","Dash"])
            .setSelection(1)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                chartView.chart.setNeedsReloadSegments()
                chartView.redraw()
            })
        )
        
    }
    
    // MARK: - Items
    
    lazy var items: NSMutableArray = {
        let items = NSMutableArray()
        for i in 0..<2 {
            if let item = createItem(atIndex: i) {
                items.add(item)
            }
        }
        return items
    }()
    
    var itemsOptionTitle: String {
        return "Polygons"
    }
    
    func createItem(atIndex index: Int) -> Any? {
        if index >= colors.count {
            return nil
        }
        let color = colors[index]
        let chart = PolygonChart()
        chart.paint?.fill?.color = color.withAlphaComponent(0.5)
        chart.paint?.stroke = nil
        return RadarChartPolygon(chart)
    }
    
    func createItemCell(withItem item: Any, atIndex index: Int)  -> UIView {
        return SectionCell()
            .setObject(item)
            .setTitle(String(format: "Polygon%ld", index))
            .setOnReload({[weak self](cell) in
                guard let self = self else {
                    return
                }
                let polygon = cell.object as! RadarChartPolygon
                polygon.needsReloadItems = true
                self.chartView.redraw()
            })
    }
    
    func itemsDidChange(_ items: NSMutableArray) {
        chartView.chart.setNeedsReloadPolygons()
        chartView.redraw()
    }
    
    // MARK: - RadarChartDataSource
    
    func numberOfIndicatorsInRadarChart(_ radarChart: RadarChart) -> Int {
        return sliderIntegerValue(forKey: "IndicatorCount", atIndex: 0)
    }
    
    func radarChart(_ radarChart: RadarChart, indicatorAtIndex index: Int) -> RadarChartIndicator {
        let indicator = RadarChartIndicator()
        setupStrokePaint(indicator.paint, Color_White, radioCellSelectionForKey("IndicatorsLine"))
        if radioCellSelectionForKey("IndicatorsText") != 0 {
            let text = ChartText()
            text.font = UIFont.systemFont(ofSize: 11)
            text.color = Color_White
            text.textOffsetByAngle = {(text, size, angle) -> CGFloat in
                return 15
            }
            text.string = String(format: "I-%ld", index)
            indicator.text = text
        }
        return indicator
    }
    
    func numberOfSegmentsInRadarChart(_ radarChart: RadarChart) -> Int {
        return sliderIntegerValue(forKey: "SegmentCount", atIndex: 0)
    }
    
    func radarChart(_ radarChart: RadarChart, segmentAtIndex index: Int) -> RadarChartSegment {
        let segment = RadarChartSegment()
        setupStrokePaint(segment.paint, Color_White, radioCellSelectionForKey("SegmentsLine"))
        segment.shape = radioCellSelectionForKey("SegmentsShape") != 0 ? .Arc : .Polygon
        segment.value = 1.0 - 1.0 / CGFloat(numberOfSegmentsInRadarChart(radarChart)) * CGFloat(index)
        return segment
    }
    
    func numberOfPolygonsInRadarChart(_ radarChart: RadarChart) -> Int {
        return items.count
    }
    
    func radarChart(_ radarChart: RadarChart, polygontAtIndex index: Int) -> RadarChartPolygon {
        return items[index] as! RadarChartPolygon
    }
    
    func radarChart(_ radarChart: RadarChart, itemOfChartForPolygonAtIndexPath indexPath: IndexPath) -> PolygonChartItem {
        let item = PolygonChartItem(CGFloat.random(in: 0...1))
        item.axisPaint = nil
        return item
    }
    
    // MARK: - Animation
    
    override func appendAnimationKeys(_ animationKeys: NSMutableArray) {
        super.appendAnimationKeys(animationKeys)
        animationKeys.add("Values")
    }
    
    override func prepareAnimationOfChartView(forKeys keys: [String]) {
        super.prepareAnimationOfChartView(forKeys: keys)
        
        if (keys.contains("Values")) {
            for polygon in items as! [RadarChartPolygon] {
                guard let items = polygon.chart.items else {
                    continue
                }
                for item in items {
                    item.transformValue = TransformCGFloat(item.value, CGFloat.random(in: 0...1))
                }
            }
        }
        
    }
    
    
    
}
